File: serialization/serialization-test-utils/src/test/java/com/cedarsoft/serialization/ComplexStaxMateSerializerTest.java
package com.cedarsoft.serialization;

import com.cedarsoft.Version;
import com.cedarsoft.VersionRange;
import com.cedarsoft.lookup.Lookup;
import com.cedarsoft.serialization.stax.AbstractStaxMateSerializer;
import com.cedarsoft.serialization.stax.AbstractStaxMateSerializerTest;
import org.codehaus.stax2.XMLStreamReader2;
import org.codehaus.staxmate.out.SMOutputElement;
import org.jetbrains.annotations.NotNull;

import javax.xml.stream.XMLStreamException;
import java.io.IOException;

import static org.testng.Assert.*;

/**
 *
 */
public class ComplexStaxMateSerializerTest extends AbstractStaxMateSerializerTest<String> {
  @NotNull
  @Override
  protected AbstractStaxMateSerializer<String> getSerializer() {
    final AbstractStaxMateSerializer<String> stringSerializer = new AbstractStaxMateSerializer<String>( "asdf", new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 0, 0 ) ) ) {
      @Override
      @NotNull
      public SMOutputElement serialize( @NotNull SMOutputElement serializeTo, @NotNull String object, @NotNull Lookup context ) throws XMLStreamException {
        serializeTo.addCharacters( object );
        return serializeTo;
      }

      @Override
      @NotNull
      public String deserialize( @NotNull XMLStreamReader2 deserializeFrom, @NotNull Lookup context ) throws XMLStreamException {
        deserializeFrom.next();
        String text = deserializeFrom.getText();
        closeTag( deserializeFrom );
        return text;
      }
    };

    return new AbstractStaxMateSerializer<String>( "aString", new VersionRange( new Version( 1, 0, 0 ), new Version( 1, 0, 0 ) ) ) {
      @Override
      @NotNull
      public SMOutputElement serialize( @NotNull SMOutputElement serializeTo, @NotNull String object, @NotNull Lookup context ) throws IOException, XMLStreamException {
        stringSerializer.serialize( serializeTo.addElement( "sub" ), object, context );
        serializeTo.addElement( "emptyChild" ).addCharacters( "" );

        return serializeTo;
      }

      @Override
      @NotNull
      public String deserialize( @NotNull XMLStreamReader2 deserializeFrom, @NotNull Lookup context ) throws IOException, XMLStreamException {
        nextTag( deserializeFrom, "sub" );
        String string = stringSerializer.deserialize( deserializeFrom, context );

        assertEquals( getChildText( deserializeFrom, "emptyChild" ), "" );
        closeTag( deserializeFrom );

        return string;
      }
    };
  }

  @NotNull
  @Override
  protected String createObjectToSerialize() {
    return "asdf";
  }

  @NotNull
  @Override
  protected String getExpectedSerializedString() {
    return "<aString><sub>asdf</sub><emptyChild/></aString>";
  }
}

File: serialization/stax-mate/src/test/kotlin/it/neckar/open/serialization/stax/mate/ComplexStaxMateSerializerTest.kt
/**
 * Copyright (C) cedarsoft GmbH.
 *
 * Licensed under the GNU General Public License version 3 (the "License")
 * with Classpath Exception; you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.cedarsoft.org/gpl3ce
 * (GPL 3 with Classpath Exception)
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 only, as
 * published by the Free Software Foundation. cedarsoft GmbH designates this
 * particular file as subject to the "Classpath" exception as provided
 * by cedarsoft GmbH in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 3 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 3 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact cedarsoft GmbH, 72810 Gomaringen, Germany,
 * or visit www.cedarsoft.com if you need additional information or
 * have any questions.
 */
package it.neckar.open.serialization.stax.mate

import it.neckar.open.serialization.StreamSerializer
import it.neckar.open.serialization.test.utils.AbstractXmlSerializerTest2
import it.neckar.open.serialization.test.utils.Entry
import it.neckar.open.version.Version
import it.neckar.open.version.VersionRange
import org.codehaus.staxmate.out.SMOutputElement
import org.junit.jupiter.api.Assertions
import javax.annotation.Nonnull
import javax.xml.stream.XMLStreamException
import javax.xml.stream.XMLStreamReader

/**
 *
 */
class ComplexStaxMateSerializerTest : AbstractXmlSerializerTest2<String>() {
  @Nonnull
  override fun getSerializer(): StreamSerializer<String> {
    val stringSerializer: AbstractStaxMateSerializer<String> = object : AbstractStaxMateSerializer<String>("asdf", "asdf", VersionRange(Version(1, 0, 0), Version(1, 0, 0))) {
      @Throws(XMLStreamException::class)
      override fun serialize(@Nonnull serializeTo: SMOutputElement, @Nonnull objectToSerialize: String, @Nonnull formatVersion: Version) {
        assert(isVersionWritable(formatVersion))
        serializeTo.addCharacters(objectToSerialize)
      }

      @Nonnull
      @Throws(XMLStreamException::class)
      override fun deserialize(@Nonnull deserializeFrom: XMLStreamReader, @Nonnull formatVersion: Version): String {
        assert(isVersionReadable(formatVersion))
        deserializeFrom.next()
        val text = deserializeFrom.text
        closeTag(deserializeFrom)
        return text
      }
    }
    return object : AbstractStaxMateSerializer<String>("aString", "asdf", VersionRange(Version(1, 0, 0), Version(1, 0, 0))) {
      @Throws(Exception::class)
      override fun serialize(@Nonnull serializeTo: SMOutputElement, @Nonnull objectToSerialize: String, @Nonnull formatVersion: Version) {
        stringSerializer.serialize(serializeTo.addElement(serializeTo.namespace, "sub"), objectToSerialize, formatVersion)
        serializeTo.addElement(serializeTo.namespace, "emptyChild").addCharacters("")
      }

      @Nonnull
      @Throws(Exception::class)
      override fun deserialize(@Nonnull deserializeFrom: XMLStreamReader, @Nonnull formatVersion: Version): String {
        assert(isVersionReadable(formatVersion))
        nextTag(deserializeFrom, "sub")
        val string = stringSerializer.deserialize(deserializeFrom, formatVersion)
        Assertions.assertEquals("", getChildText(deserializeFrom, "emptyChild"))
        closeTag(deserializeFrom)
        return string
      }
    }
  }

  companion object {
    val ENTRY1: Entry<out String> = create("asdf", "<aString><sub>asdf</sub><emptyChild/></aString>")
  }
}
